README.md

```diff
@@ -1,4 +1,134 @@
+# Validator Smart Contracts
 
-This repository contains Ethereum smart contracts that can be used to select and change the QBFT validators for Besu and GoQuorum Ethereum clients.
+The QBFT consensus protocol implementation in the Quorum clients (GoQuorum and Besu) allows users to use a smart
+contract to specify the validators that are used to propose and vaidate blocks.
 
+These smart contracts need to implement the interface contract ValidatorSmartContractInterface.sol, specifically the 
+function  
+
+    function getValidators() external view returns (address[] memory)
+
+This repository contains a simple implementation of this interface.
+
+General information about QBFT can be found here: https://besu.hyperledger.org/en/stable/HowTo/Configure/Consensus-Protocols/QBFT/
+
+## Allowlist Based Contract
+
+This smart contract can be found in the directory _contracts/allowlist_. 
+
+The contract holds a list of accounts (the allowlist) that are each allowed to nominate one QBFT validator. Each of 
+these accounts can use transactions to call certain functions on the contract to activate and deactivate a validator. 
+These accounts can also vote to add additional accounts to the allowlist, and to remove them.
+
+To make use of this contract starting from the genesis block, the genesis file needs to contain this contract, as well 
+as the storage content for the initial allowed accounts and validators. See 
+[chapter "Genesis File Content"](#genesis-file-content) for a utility to generate the required storage. 
+
+**See the web3-js based script in _scripts/allowlist/cli_ for a simple cli script to call the allowlist smart contract functions.**
+
+Accounts on the allowlist can use the API of this contract to
+* activate a validator
+* deactivate a validator
+* vote to add an account to the allowlist
+* vote to remove an account from the allowlist
+* remove votes they have cast to add or remove an account
+* execute the vote count for an account to be added or removed
+
+For an election to be successful more than 50% of the current members of the allowlist have to vote.
+
+To get the votes counted and (if successful) the outcome enacted (the specified account added or removed),
+the function _countVotes_ needs to be called.
+
+A number of events are emitted to enable users to get information about changes to the validators, allowed accounts,
+and voting.
+
+To be able to use this contract starting from the genesis block of a blockchain the initial state of this contract
+needs to be set in the genesis file. The `scripts/allowlist/genesisContent` directory of this
+repository contains a javascript script that creates the storage section for this contract.
+
+### Compile the Contract
+
+To create the contract code that needs to be specified in the genesis file, this contract needs to be compiled with the
+option `--bin-runtime`.
+
+After copying the interface contract into this directory
+
+    cp ../ValidatorSmartContractInterface.sol .
```

Comment:
does it need to be in a separate directory?

Reply:
The problem is that you need to allow solidity to access files other than the current directory and it's subdirectories. That makes the command line quite ugly ....